From d0896c071e7bf35945f391bcf6cc07845117baa3 Mon Sep 17 00:00:00 2001
From: Wei-Luan Wang <weiluanwang@google.com>
Date: Fri, 26 Dec 2025 00:55:32 +0000
Subject: [PATCH] Add base::to_underlying before C++23

std::to_underlying is not available for C++20.

Revert "Delete unused base/types/cxx23_to_underlying.h"
This reverts commit ef6e4802342840470b667df82e81af4c68819d4c.

Revert "Migrate usages of base::to_underlying in //base"
This reverts commit 86dec152c75a4a3f01802e7f57dad3866badd5cd.
- Changes in base/task/thread_pool/priority_queue.{h|cc} are not needed
  after crrev.com/c/7545201.

Revert "Make base::to_underlying an alias to std::to_underlying"
This reverts commit 23c18f85275491a21335d2a260a8379c84cf50b5.

Change-Id: I450dd0370048ac435c740e7a49193aec2e130128
---
 base/apple/mach_port_rendezvous.cc            |  1 +
 base/check.cc                                 |  6 ++--
 base/containers/enum_set.h                    | 31 ++++++++--------
 base/functional/callback_internal.cc          |  5 ++-
 base/metrics/histogram_macros_internal.h      |  4 +--
 base/types/cxx23_to_underlying.h              | 23 ++++++++++++
 base/types/cxx23_to_underlying_unittest.cc    | 35 +++++++++++++++++++
 .../c++/basic_struct.test-mojom.cc.golden     |  1 +
 .../c++/basic_union.test-mojom.cc.golden      |  1 +
 .../c++/interface.test-mojom.cc.golden        |  5 +--
 .../optional_primitives.test-mojom.cc.golden  |  1 +
 .../c++/results.test-mojom.cc.golden          | 11 +++---
 .../c++/typemap.test-mojom.cc.golden          |  5 +--
 .../bindings/tests/sync_method_unittest.cc    |  7 ++--
 .../cpp_templates/interface_definition.tmpl   |  8 ++---
 .../generators/cpp_templates/module.cc.tmpl   |  1 +
 16 files changed, 106 insertions(+), 39 deletions(-)
 create mode 100644 base/types/cxx23_to_underlying.h
 create mode 100644 base/types/cxx23_to_underlying_unittest.cc

diff --git a/base/apple/mach_port_rendezvous.cc b/base/apple/mach_port_rendezvous.cc
index 6827382152..7cd6db6ca0 100644
--- a/base/apple/mach_port_rendezvous.cc
+++ b/base/apple/mach_port_rendezvous.cc
@@ -21,6 +21,7 @@
 #include "base/notreached.h"
 #include "base/numerics/byte_conversions.h"
 #include "base/strings/stringprintf.h"
+#include "base/types/cxx23_to_underlying.h"
 
 namespace base {
 
diff --git a/base/check.cc b/base/check.cc
index 2ce900f0f4..d69a6aa13d 100644
--- a/base/check.cc
+++ b/base/check.cc
@@ -5,7 +5,6 @@
 #include "base/check.h"
 
 #include <optional>
-#include <utility>
 
 #include "base/check_op.h"
 #include "base/check_version_internal.h"
@@ -14,6 +13,7 @@
 #include "base/debug/dump_without_crashing.h"
 #include "base/logging.h"
 #include "base/thread_annotations.h"
+#include "base/types/cxx23_to_underlying.h"
 #include "build/build_config.h"
 
 namespace logging {
@@ -67,7 +67,7 @@ LogSeverity GetDumpSeverity() {
 
 LogSeverity GetNotFatalUntilSeverity(base::NotFatalUntil fatal_milestone) {
   if (fatal_milestone != base::NotFatalUntil::NoSpecifiedMilestoneInternal &&
-      std::to_underlying(fatal_milestone) <= BASE_CHECK_VERSION_INTERNAL) {
+      base::to_underlying(fatal_milestone) <= BASE_CHECK_VERSION_INTERNAL) {
     return LOGGING_FATAL;
   }
   return GetDumpSeverity();
@@ -112,7 +112,7 @@ void MaybeSetFatalMilestoneCrashKey(base::NotFatalUntil fatal_milestone) {
   }
   base::debug::SetCrashKeyString(
       GetFatalMilestoneCrashKey(),
-      base::NumberToString(std::to_underlying(fatal_milestone)));
+      base::NumberToString(base::to_underlying(fatal_milestone)));
 }
 
 void DumpWithoutCrashing(base::debug::CrashKeyString* message_key,
diff --git a/base/containers/enum_set.h b/base/containers/enum_set.h
index 4cf3168c6b..cc3a1b4645 100644
--- a/base/containers/enum_set.h
+++ b/base/containers/enum_set.h
@@ -17,6 +17,7 @@
 #include "base/check.h"
 #include "base/check_op.h"
 #include "base/memory/raw_ptr.h"
+#include "base/types/cxx23_to_underlying.h"
 
 namespace base {
 
@@ -61,7 +62,7 @@ class EnumSet {
   static constexpr E kMinValue = MinEnumValue;
   static constexpr E kMaxValue = MaxEnumValue;
   static constexpr size_t kValueCount =
-      std::to_underlying(kMaxValue) - std::to_underlying(kMinValue) + 1;
+      to_underlying(kMaxValue) - to_underlying(kMinValue) + 1;
 
   static_assert(kMinValue <= kMaxValue,
                 "min value must be no greater than max value");
@@ -211,37 +212,37 @@ class EnumSet {
 
   // Returns an EnumSet constructed from |bitmask|.
   static constexpr EnumSet FromEnumBitmask(const uint64_t bitmask) {
-    static_assert(std::to_underlying(kMaxValue) < 64,
+    static_assert(to_underlying(kMaxValue) < 64,
                   "The highest enum value must be < 64 for FromEnumBitmask ");
-    static_assert(std::to_underlying(kMinValue) >= 0,
+    static_assert(to_underlying(kMinValue) >= 0,
                   "The lowest enum value must be >= 0 for FromEnumBitmask ");
-    return EnumSet(EnumBitSet(bitmask >> std::to_underlying(kMinValue)));
+    return EnumSet(EnumBitSet(bitmask >> to_underlying(kMinValue)));
   }
   // Returns a bitmask for the EnumSet.
   constexpr uint64_t ToEnumBitmask() const {
-    static_assert(std::to_underlying(kMaxValue) < 64,
+    static_assert(to_underlying(kMaxValue) < 64,
                   "The highest enum value must be < 64 for ToEnumBitmask ");
-    static_assert(std::to_underlying(kMinValue) >= 0,
+    static_assert(to_underlying(kMinValue) >= 0,
                   "The lowest enum value must be >= 0 for FromEnumBitmask ");
-    return enums_.to_ullong() << std::to_underlying(kMinValue);
+    return enums_.to_ullong() << to_underlying(kMinValue);
   }
 
   // Returns a uint64_t bit mask representing the values within the range
   // [64*n, 64*n + 63] of the EnumSet.
   constexpr std::optional<uint64_t> GetNth64bitWordBitmask(size_t n) const {
     // If the EnumSet contains less than n 64-bit masks, return std::nullopt.
-    if (std::to_underlying(kMaxValue) / 64 < n) {
+    if (to_underlying(kMaxValue) / 64 < n) {
       return std::nullopt;
     }
 
     std::bitset<kValueCount> mask = ~uint64_t{0};
     std::bitset<kValueCount> bits = enums_;
-    if (std::to_underlying(kMinValue) < n * 64) {
-      bits >>= n * 64 - std::to_underlying(kMinValue);
+    if (to_underlying(kMinValue) < n * 64) {
+      bits >>= n * 64 - to_underlying(kMinValue);
     }
     uint64_t result = (bits & mask).to_ullong();
-    if (std::to_underlying(kMinValue) > n * 64) {
-      result <<= std::to_underlying(kMinValue) - n * 64;
+    if (to_underlying(kMinValue) > n * 64) {
+      result <<= to_underlying(kMinValue) - n * 64;
     }
     return result;
   }
@@ -353,13 +354,13 @@ class EnumSet {
   // Converts a value to/from an index into |enums_|.
   static constexpr size_t ToIndex(E value) {
     CHECK(InRange(value));
-    return static_cast<size_t>(std::to_underlying(value)) -
-           static_cast<size_t>(std::to_underlying(MinEnumValue));
+    return static_cast<size_t>(to_underlying(value)) -
+           static_cast<size_t>(to_underlying(MinEnumValue));
   }
 
   static constexpr E FromIndex(size_t i) {
     DCHECK_LT(i, kValueCount);
-    return static_cast<E>(std::to_underlying(MinEnumValue) + i);
+    return static_cast<E>(to_underlying(MinEnumValue) + i);
   }
 
   EnumBitSet enums_;
diff --git a/base/functional/callback_internal.cc b/base/functional/callback_internal.cc
index de55b8c340..27f6900728 100644
--- a/base/functional/callback_internal.cc
+++ b/base/functional/callback_internal.cc
@@ -4,10 +4,9 @@
 
 #include "base/functional/callback_internal.h"
 
-#include <utility>
-
 #include "base/check.h"
 #include "base/notreached.h"
+#include "base/types/cxx23_to_underlying.h"
 
 namespace base::internal {
 
@@ -18,7 +17,7 @@ bool QueryCancellationTraitsForNonCancellables(
     BindStateBase::CancellationQueryMode mode) {
   // Non-cancellables are never cancelled and always valid, which means the
   // response for each mode is the same as its underlying value.
-  return std::to_underlying(mode);
+  return to_underlying(mode);
 }
 
 }  // namespace
diff --git a/base/metrics/histogram_macros_internal.h b/base/metrics/histogram_macros_internal.h
index 0a22535e60..09a9458fac 100644
--- a/base/metrics/histogram_macros_internal.h
+++ b/base/metrics/histogram_macros_internal.h
@@ -11,13 +11,13 @@
 #include <limits>
 #include <memory>
 #include <type_traits>
-#include <utility>
 
 #include "base/dcheck_is_on.h"
 #include "base/metrics/histogram.h"
 #include "base/metrics/sparse_histogram.h"
 #include "base/no_destructor.h"
 #include "base/time/time.h"
+#include "base/types/cxx23_to_underlying.h"
 
 // This is for macros and helpers internal to base/metrics. They should not be
 // used outside of this directory. For writing to UMA histograms, see
@@ -33,7 +33,7 @@ struct EnumSizeTraits {
     if constexpr (requires { Enum::kMaxValue; }) {
       // Since the UMA histogram macros expect a value one larger than the max
       // defined enumerator value, add one.
-      return static_cast<uintmax_t>(std::to_underlying(Enum::kMaxValue) + 1);
+      return static_cast<uintmax_t>(base::to_underlying(Enum::kMaxValue) + 1);
     } else {
       static_assert(
           sizeof(Enum) == 0,
diff --git a/base/types/cxx23_to_underlying.h b/base/types/cxx23_to_underlying.h
new file mode 100644
index 0000000000..8250bdd125
--- /dev/null
+++ b/base/types/cxx23_to_underlying.h
@@ -0,0 +1,23 @@
+// Copyright 2022 The Chromium Authors
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef BASE_TYPES_CXX23_TO_UNDERLYING_H_
+#define BASE_TYPES_CXX23_TO_UNDERLYING_H_
+
+#include <type_traits>
+
+namespace base {
+
+// Implementation of C++23's std::to_underlying.
+//
+// Reference: https://en.cppreference.com/w/cpp/utility/to_underlying
+template <typename EnumT>
+  requires(std::is_enum_v<EnumT>)
+constexpr std::underlying_type_t<EnumT> to_underlying(EnumT e) noexcept {
+  return static_cast<std::underlying_type_t<EnumT>>(e);
+}
+
+}  // namespace base
+
+#endif  // BASE_TYPES_CXX23_TO_UNDERLYING_H_
diff --git a/base/types/cxx23_to_underlying_unittest.cc b/base/types/cxx23_to_underlying_unittest.cc
new file mode 100644
index 0000000000..d0c507f6dd
--- /dev/null
+++ b/base/types/cxx23_to_underlying_unittest.cc
@@ -0,0 +1,35 @@
+// Copyright 2022 The Chromium Authors
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "base/types/cxx23_to_underlying.h"
+
+#include "testing/gtest/include/gtest/gtest.h"
+
+namespace base {
+
+TEST(Cxx23ToUnderlying, Basic) {
+  enum Enum : int {
+    kOne = 1,
+    kTwo = 2,
+  };
+
+  enum class ScopedEnum : char {
+    kOne = 1,
+    kTwo = 2,
+  };
+
+  static_assert(std::is_same_v<decltype(to_underlying(kOne)), int>, "");
+  static_assert(std::is_same_v<decltype(to_underlying(kTwo)), int>, "");
+  static_assert(to_underlying(kOne) == 1, "");
+  static_assert(to_underlying(kTwo) == 2, "");
+
+  static_assert(std::is_same_v<decltype(to_underlying(ScopedEnum::kOne)), char>,
+                "");
+  static_assert(std::is_same_v<decltype(to_underlying(ScopedEnum::kTwo)), char>,
+                "");
+  static_assert(to_underlying(ScopedEnum::kOne) == 1, "");
+  static_assert(to_underlying(ScopedEnum::kTwo) == 2, "");
+}
+
+}  // namespace base
